x-pack/legacy/plugins/code/server/distributed/cluster/cluster_service.ts

```diff
@@ -0,0 +1,104 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import * as lodash from 'lodash';
+import { ClusterMeta } from './cluster_meta';
+import { EsClient } from '../../lib/esqueue';
+import { RepositoryObjectClient } from '../../search';
+import { Poller } from '../../poller';
+import { Logger } from '../../log';
+import { Repository } from '../../../model';
+import { RoutingTable } from './routing_table';
+import { ClusterState } from './cluster_state';
+import { ClusterStateEvent } from './cluster_state_event';
+
+/**
+ * ClusterService synchronizes the core cluster states with the remote, and provides methods to read the current state,
+ * and to register listeners to watch the state change.
+ */
+export class ClusterService {
+  private readonly clusterStateListeners: ClusterStateListener[] = [];
+  private readonly clusterStatePoller = new Poller<void>({
+    functionToPoll: () => {
+      return this.pollClusterState();
+    },
+    pollFrequencyInMillis: 5000,
+    trailing: true,
+  });
+
+  private currentState: ClusterState = new ClusterState(new ClusterMeta([]), new RoutingTable([]));
+
+  constructor(
+    public readonly localNodeId: string,
+    public readonly esClient: EsClient,
+    public readonly logger: Logger,
+    private readonly repositoryObjectClient: RepositoryObjectClient = new RepositoryObjectClient(
+      esClient
+    )
+  ) {}
+
+  public async start() {
+    this.clusterStatePoller.start();
+  }
+
+  public async stop() {
+    this.clusterStatePoller.stop();
+  }
+
+  /**
+   * Sync the cluster state with the remote storage.
+   *
+   * For now, in a multi-node cluster, where there is only a single stateful node, we only need to get repository objects,
+   * and assign all the repository to the code (local) node in the (virtual) routing table.
+   */
+  async pollClusterState(): Promise<void> {
```

Comment:
I know you are relying on polling to sync with es regarding all the repositories, but do you think it makes sense to immediately update the clusterstate when some repository has been removed or created? 

I am thinking of the case when a repository clone job has been enqueued, this will add another 5 seconds delay for the routing table to be able to register the new repository for the esqueue successfully dequeue this clone job. Similar issue for repository delete job.

Reply:
Also, since we are snapshotting a routing table in the memory here while its source of truth lies in ES eventually, having this 5s polling will add another layer of consistency sync between them. This will expose more cases when inconsistency could happen. Another option might be looking up the routing table in the ES on the fly for each routing table related requests to avoid the consistency here, but of course, it's a bit costly.

Reply:
Yes, the current implementation may increase the latency, and it can be optimized as you suggested, and someday we may reduce the latency to milliseconds level if ES provides us some API to subscribe changes of shard data, like those used in CCR.
The consistency problem is tricky, since our implementation is asynchronous in general, there may always be inconsistency even we look up the routing table in the ES on the fly every time, so the key point may be to know where and what kind of inconsistency is acceptable. We may have further discussion on this topic.